prow/crier/controller_test.go

```diff
@@ -0,0 +1,244 @@
+	return "triggerReporter"
+}
+
+func (f triggerReporter) ShouldReport(pj *prowv1.ProwJob) bool {
+	return pj.Status.State == prowv1.TriggeredState
+}
+
+// Fake Informer
+// Sets: The Prow Job Test Cases
+// Asserts: The Controller registers for the queue to be populated
+type fakeInformer struct {
+	eventChannel chan string
+}
+
+func (f fakeInformer) Get(name string) (*prowv1.ProwJob, error) {
+	jobs := map[string]*prowv1.ProwJob{
+		"TestJob1": {
+			Spec: prowv1.ProwJobSpec{
+				Job: "TestJob1",
+			},
+			Status: prowv1.ProwJobStatus{
+				State: prowv1.FailureState,
+			},
+		},
+		"TestJob2": {
+			Spec: prowv1.ProwJobSpec{
+				Job: "TestJob2",
+			},
+			Status: prowv1.ProwJobStatus{
+				State: prowv1.TriggeredState,
+			},
+		},
+	}
+
+	pj, found := jobs[name]
+	if !found {
+		return nil, errors.New("Can't Find ProwJob")
+	}
+	return pj, nil
+}
+
+func (f fakeInformer) ProwJobs(namespace string) prowLister.ProwJobNamespaceLister {
+	return f
+}
+
+func (f fakeInformer) Informer() cache.SharedIndexInformer {
+	return f
+}
+
+func (f fakeInformer) Lister() prowLister.ProwJobLister {
+	return f
+}
+
+func (f fakeInformer) AddEventHandler(handler cache.ResourceEventHandler) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Registered"
+	}
+}
+
+func (f fakeInformer) Run(stopCh <-chan struct{}) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Run"
+	}
+}
+
+func (f fakeInformer) AddEventHandlerWithResyncPeriod(handler cache.ResourceEventHandler, resyncPeriod time.Duration) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Registered"
+	}
+}
+
+func (fakeInformer) HasSynced() bool {
+	return true
+}
+
+func (fakeInformer) GetStore() cache.Store                                            { return nil }
+func (fakeInformer) GetController() cache.Controller                                  { return nil }
+func (fakeInformer) LastSyncResourceVersion() string                                  { return "" }
+func (fakeInformer) AddIndexers(indexers cache.Indexers) error                        { return nil }
+func (fakeInformer) GetIndexer() cache.Indexer                                        { return nil }
+func (fakeInformer) List(selector labels.Selector) (ret []*prowv1.ProwJob, err error) { return nil, nil }
+
+func TestController_processNextItem(t *testing.T) {
+	tests := []struct {
+		name         string
+		expectReport bool
+	}{
+		{
+			name:         "TestJob1",
+			expectReport: false,
+		},
+		{
+			name:         "TestJob2",
+			expectReport: true,
+		},
+		{
+			name:         "DoesNotExist",
+			expectReport: false,
+		},
+	}
+
+	q := kube.RateLimiter(controllerName)
+	for _, test := range tests {
+		q.Add(test.name)
+	}
+
+	var rp triggerReporter
+	var isReported bool
+	rp.reportedItem = &isReported
+
+	cs := fake.NewSimpleClientset()
+	var inf fakeInformer
+	nmwrk := 1
+	wg := &sync.WaitGroup{}
+	c := NewController(cs, q, inf, rp, nmwrk, wg)
+
+	for _, test := range tests {
+		isReported = false
```

Comment:
This local variable is confusing. Seems like you can dispense with it and just read rp.reportedItem

Reply:
Because the function signature of Report passes triggerReporter by value, any changes made to it by the Report function won't stick. That's why rp.reportedItem is a *bool, not a bool; it's actually pointing to "isReported".

I'm not certain that referring to the local variable as "*rp.reportedItem" more clearly indicates what's going on here, but I've made the change. Let me know what you think about its readability.

Reply:
Change `func (f triggerReporter) Report` to `func (f *triggerReporter) Report`